Policy simplification

tgraf opened the pull request:
Fixes: #674

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+// accumulated additively
+//
+// The rule is split into an ingress section which contains all rules
+// applicable at ingress, and an egress section applicable at egress.
```
applicable -> enforced

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+// accumulated additively
+//
+// The rule is split into an ingress section which contains all rules
+// applicable at ingress, and an egress section applicable at egress.
```
I would also specify that for a flow between two endpoints to be allowed by Cilium, it must be allowed by both the egress section of the source endpoint and by the ingress section of the destination endpoint.

tgraf replied:
That is only true to some extent. Endpoint to endpoint policy is only specified at Ingress. We only do egress specification for things where we do not control the receiving endpoint.

I've added a note to clarify this.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// rule. Rules cannot be identified by comment.
+	//
+	// +optional
+	Comment string `json:"comment,omitempty"`
```
Comment -> Description (?)

tgraf replied:
I inherited comment from iptables but I like Description better as well, fixed.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// If omitted, no rules will be applied at ingress via this rule.
+	//
+	// +optional
+	Ingress *IngressRule `json:"ingress,omitempty"`
```
This should be a slice, not a single rule.

tgraf replied:
All members of IngressRule are sliced which saves us to enforce oneOf at IngressRule level. What benefit do you see in one over the over? I'm fine with both actually.

rlenglet replied:
Because the semantics is different.
I would like to be able to specify:
- allow from 10/8 to port 80
and
- allow from 11/8 to port 443

That's very different from:
- allow from "either 10/8 or 11/8" to "either port 80 or port 443"

tgraf replied:
As mentioned in another comment as well: The datapath does support the coupling of L3 and L4 rules at this point and we have to support the decoupled version anyway for k8s policy.

I'm fine changing this as I said but we can't support the coupling right now.

rlenglet replied:
Actually you could. It would just cost you the calculation of a cartesian product of the k8s L3 rules with the K8s L4 rules in the same k8s policy, to obtain those "coupled" rule versions. It may be costly only in case of complicated k8s policies, e.g. with contain many criteria. I don't think that will be common, nor very costly.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// If omitted, no rules will be applied at egress via this rule
+	//
+	// +optional
+	Egress *EgressRule `json:"egress,omitempty"`
```
This should be a slice, not a single rule.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// endpoints which are whitelisted in FromEndpoints.
+	//
+	// +optional
+	FromCIDR []CIDR `json:"toCIDR,omitempty"`
```
toCIDR -> fromCIDR (in the json tag)

tgraf replied:
Fixed

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+// IngressRule contains all rule types which can be applied at ingress,
+// i.e. network traffic that originates outside of the endpoint and
+// is entering the endpoint selected by the endpointSelector.
+//
```
I would clarify that each rule is conceptually a conjunction of predicates (one predicate per struct member): all specified predicates must match for the rule to apply. If a given predicate is nil, it is "wildcarded", i.e. it matches any traffic.

tgraf replied:
What you say is currently not true. The `FromEndpoints` is in no way correlated to the `ToPorts`. We can possibly support that in the future but it is a substantial change. It also means that we can no longer map a single k8s `NetworkPolicy` to a single Cilium rule which would complicate usage quite a bit.

tgraf replied:
The last sentence in my last comment is actually inaccurate, if we move to a slice of Ingress/EgressRule we could still support k8s NetworkPolicy in a single Rule, split over multiple IngressRules.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// rule is allowed to receive connections from. This will match on the
+	// source IP address of incoming connections.
+	//
+	// This restriction does not apply to connections from endpoints which
```
I find this confusing. I would prefer to define each rule as a conjunction of predicates.
I would personally find it clearer if FromEndpoint and FromCIDR could not be used simultaneously. For instance, by specifying this as an anyOf in JSON Schema.

tgraf replied:
Even if we enforce an anyOf in this structure, a user can still inject two rules with an identical endpoint selector containing a FromEndpoint and FromCIDR each. We can't enforce this from happening. `FromCIDR` is in addition to `FromEndpoints` which is very simple to understand. If we enforce an anyOf here it creates an impression of something that is not enforced over the entire policy.

rlenglet replied:
I don't see the problem here. IMHO the set of rules in a policy should form a union ("or") of the rules. So if you have 2 rules:
- rule A: FromEndpoint: blah
- rule B: FromCIDR: 10/8

Then I would expect traffic to be allowed if EITHER the source has label "blah" OR it has an IP address in prefix 10/8.
Why would that create any problem?

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// protocol which the endpoint subject to the rule is allowed to
+	// connect to.
+	//
+	// If omitted and no other IngressRule applies a ToPorts rule on the
```
Why mentioning "IngressRule" here? Did you mean "EgressRule"?

tgraf replied:
I meant EgressRule but I believe the paragraph is not required at all for understanding.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// protocol which the endpoint subject to the rule is allowed to
+	// connect to.
+	//
+	// If omitted and no other IngressRule applies a ToPorts rule on the
```
I find this confusing. This looks like a conjunction of rules.
I would prefer if all the rules form a disjunction, i.e. if a rule allows some flows, no addition or removal of any other rule can alter that.

tgraf replied:
Not sure which paragraph this comment refers to.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	foo string
+}
+
+// PortProtocol specifices an L4 port with an optional protocol
```
specifices -> specifies
protocol -> transport protocol

tgraf replied:
fixed

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+type Rule struct {
+	// EndpointSelector selects all endpoints which should be subject to
+	// this rule. Cannot be empty.
+	EndpointSelector EndpointSelector `json:"selector"`
```
Please consider renaming this to AppliedEndpoints. 

// Mandatory field. Identifies the endpoints where this rule should be applied.
AppliedEndpoints EndpointSelector `json:"selector"`

tgraf replied:
I like selector because it is in line with the k8s API which uses the word selector for everything to select objects to apply rules onto.

ashwinp replied:
I see. Thanks.

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// If omitted, no rules will be applied at ingress via this rule.
+	//
+	// +optional
+	Ingress *IngressRule `json:"ingress,omitempty"`
```
The field is a pointer to IngressRule, and can hold only 1 ingress rule. The document needs to be fixed, or we need to have a collection of IngressRules here.

tgraf replied:
comment fixed

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+
+// PortProtocol specifices an L4 port with an optional protocol
+//
+// TODO: allow port range?
```
It's usually useful. Actually, a list combining individual ports and ranges seems common, e.g.:
["80", "443", "2000-2100"]
The only negative aspect about it is that it requires the data to be passed as strings, but that's a small price to pay.

tgraf replied:
I agree, I will change it to a string for now so we can easily add ranges later on.

aanm replied:
What about:
```
type PortProtocol struct {
      Port PortRange
}

type PortRange stuct{
      min uint16
      max uint16
}

func NewPort(p uint16) *PortRange{
     return &PortRange{min: p, max: p}
}

func NewPortRange(min, max uint16) *PortRange{
     return &PortRange{min: min, max: max}
}

func (p *PortRange) Contains(p uint16) bool {
   return p.min >= p && p <= p.max
}
```
At least we don't deal with strings

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	Port uint16 `json:"port"`
+
+	// Protocol is the L4 protocol. If omitted, any protocol matches
+	// Accepted values: "tcp", "udp", ""/"any"
```
In case of ICMP, how can one match on the code / type?
If that's not supported, specify it.

tgraf replied:
Note added, we do not support matching on ICMP.

aanm replied:
What about port 0 with proto any?

tgraf replied:
Such a rule would never match right now. I've added the case to the `Validate()` function.

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+//
+// Either ingres, egress, or both can be provided. If both ingress and egress
+// are omitted, the rule has no effect.
+type Rule struct {
```
Please consider renaming this to Policy. A Policy can contain multiple rules. Each rule can be of type Egress or Ingress.

tgraf replied:
That's what it is. The policy is not part of the API. Cilium owns a policy repository which consists of a list of rules which can be modified via the API by adding/removing/changing rules.

ashwinp replied:
Got it. Thanks.

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// If omitted, no rules will be applied at egress via this rule
+	//
+	// +optional
+	Egress *EgressRule `json:"egress,omitempty"`
```
Can we simplify this by having a collection of Rules (a new type that only contains match predicates)? A Rule can either be of type Egress or Ingress. The current type Rule can be renamed to Policy.

tgraf replied:
See the comment from @rlenglet. I'm trying to avoid oneOf as it can't be described and enforced with swagger.

ashwinp replied:
OK, thanks.

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// endpoints which are whitelisted in fromCIDR.
+	//
+	// +optional
+	FromEndpoints []EndpointSelector `json:"fromEndpoints,omitempty"`
```
Please consider renaming this to SourceEndpoints

tgraf replied:
Source has some implications as it could be confused for source addresses. "From" is clearer in that sense, "rule selects b and allows ingress from a"

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// EndpointSelector which are allowed to communicate with the endpoint
+	// subject to the rule.
+	//
+	// If omitted and no other IngressRule applies a FromEndpoints rule on
```
Please remove this paragraph. We're referring to fromCIDR, which is defined later, and the comment is redundant as the params are tagged as optional.

tgraf replied:
Agreed, only obfuscates, removed both occurrences.

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// +optional
+	FromEndpoints []EndpointSelector `json:"fromEndpoints,omitempty"`
+
+	// FromRequires is a selector which must match in combination with at
```
I did not understand the purpose of FromRequires. The example about group=A talks about labels, but isn't this field of type endpointselector?

tgraf replied:
I have reworded the paragraph to make it clearer

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// ports.
+	//
+	// +optional
+	ToPorts []PortRule `json:"toPorts,omitempty"`
```
Please consider renaming to destinationPorts

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+	// endpoints which are whitelisted in FromEndpoints.
+	//
+	// +optional
+	FromCIDR []CIDR `json:"toCIDR,omitempty"`
```
Please consider renaming to SourceCIDRs

tgraf replied:
SourceCIDRs would definitely be mistaken for source address CIDR

tgraf replied:
2nd thought, this is exactly what we want so I'm making this change.

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,234 @@
+// TODO: allow port range?
+type PortProtocol struct {
+	// Port is an L4 port number
+	Port uint16 `json:"port"`
```
Can we make this:
Ports []uint16 ?
That way, one can specify TCP: [80, 8080], as an example.

tgraf replied:
See @rlenglet comment to move to strong to support ranges, I like that approach. We could eventually support all of `"10", "10, 11", "10-15"`

ashwinp reviewed (changes requested):
Initial review comments.

aanm commented on pkg/labels/array.go:
```diff
@@ -17,6 +17,15 @@ package labels
+func ParseLabelArray(labels ...string) LabelArray {
+	array := make([]*Label, len(labels))
+	for i := range labels {
+		array[i] = ParseLabel(labels[i])
```
What happens if a label is not valid?

tgraf replied:
How can a label not be valid?

aanm replied:
Good question, since it returns a pointer I was assuming it could be `nil`

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,228 @@
+	EndpointSelector EndpointSelector `json:"endpointSelector"`
+
+	// Ingress is a list of IngressRule which are enforced at ingress.
+	// If omitted, this rule does not apply at ingress.
```
For the `omitted` ones please also add the case where they are empty. "If omitted, or empty,..."

tgraf replied:
Agreed, will do.

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,228 @@
+	// unique, multiple rules can have overlapping or identical labels.
+	//
+	// +optional
+	Labels labels.LabelArray `json:"group,omitempty"`
```
I've been thinking about this. What about "Metadata" instead, seems wrong?
Also, why the json variable is called Group? I mean, I would like to see `Group` instead of `Labels` in the go struct.

aanm commented on pkg/policy/api/selector.go:
```diff
@@ -0,0 +1,33 @@
+// ParseEndpointSelector parses a list of labels in the format of
+// strings and returns an EndpointSelector
+func ParseEndpointSelector(list ...string) EndpointSelector {
+	array := make([]*labels.Label, len(list))
```
`[]*labels.Label` -> `labels.LabelArray`

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,228 @@
+// is entering the endpoint selected by the endpointSelector.
+//
+// - All members of this structure are optional.
+// - All rule types are considered independently
```
Does it this means all slices in this structure will be "ORed" or "ANDed"?

tgraf replied:
ORed. Rephrased.

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,228 @@
+	// connections from 10.3.9.1
+	//
+	// +optional
+	FromCIDR []CIDR `json:"fromCIDR,omitempty"`
```
`[]net.IPNet`, maybe?

tgraf replied:
I've changed the CIDR struct to contain a string, that gives us the most flexibility going forward

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,228 @@
+	//
+	// Example:
+	// Any endpoint with the label "role=frontend" is allowed to initiate
+	// connections on port 8080/tcp
```
s/`on port`/`to port`?

tgraf replied:
Fixed

aanm commented on cilium/cmd/policy.go:
```diff
@@ -142,89 +132,86 @@ func loadPolicyFile(path string) (*policy.Node, error) {
-			}
-		} else {
-			*node = p
+			return []api.Rule{}, err
```
Small nit, you can `return nil, err`

tgraf replied:
Fixed

aanm commented on daemon/policy.go:
```diff
@@ -229,7 +234,8 @@ func NewDeletePolicyPathHandler(d *Daemon) DeletePolicyPathHandler {
 func (h *deletePolicyPath) Handle(params DeletePolicyPathParams) middleware.Responder {
 	d := h.daemon
-	if err := d.PolicyDelete(params.Path, ""); err != nil {
+	// FIXME
```
What to fix about here?

aanm commented on daemon/policy.go:
```diff
@@ -247,16 +253,17 @@ func NewPutPolicyPathHandler(d *Daemon) PutPolicyPathHandler {
 	}
 
-	return NewPutPolicyPathOK().WithPayload(models.PolicyTree(node.JSONMarshal()))
+	json := policy.JSONMarshalRules(rules)
```
Maybe we should rename this to `RuleToString`? Otherwise some could just used `json.Marshal(rules)`

aanm commented on daemon/policy_test.go:
```diff
@@ -38,68 +36,47 @@ var (
-					{
-						Action: api.ACCEPT,
-						Labels: labels.LabelArray{lblPete},
+	lblProd := labels.ParseLabel("Prod")
```
What happens with the label source?

tgraf replied:
It defaults to cilium

aanm commented on pkg/policy/repository.go:
```diff
@@ -0,0 +1,198 @@
+		switch r.canReach(ctx) {
+		// The rule contained a constraint which was not met, this
+		// connection is not allowed
+		case api.Denied:
```
I don't get this, if the policy has multiple rules, and one of the rules can't be reachable why does that means it is automatically denied?

Shouldn't you first check for all the rules that have its ingress matching the `ctx.To` and then check if one of them allows receiving from `ctx.From`?

tgraf replied:
`api.Denied` is only returned if it is denied for sure. This only happens for Require rules. If the requirement for a require rule is not met, the potential consumer is never allowed.

aanm commented on pkg/policy/repository.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+// AddList inserts a rule into the policy repository
+func (p *Repository) AddList(rules []*api.Rule) error {
```
Why a pointer to the rule itself?

tgraf replied:
I've masked it behind a new `api.Rules` type now so we can easily change this. I don't see a problem with a pointer.

aanm commented on pkg/policy/repository.go:
```diff
@@ -0,0 +1,198 @@
+	defer p.Mutex.Unlock()
+
+	// Validate entire rule list first and only append array if
+	// all rules are valid
```
:+1:

aanm commented on pkg/policy/repository.go:
```diff
@@ -0,0 +1,198 @@
+
+// GetJSON returns all rules of the policy repository as string in JSON
+// representation
+func (p *Repository) GetJSON() string {
```
~~Why not implement the `MarshalJSON` itself?~~ because it returns byte and error

Why not `ToString()` or `String()` or even `GoString()`?

tgraf replied:
Because we return JSON :-) It's not meant to be human readable.

aanm commented on pkg/policy/rule.go:
```diff
@@ -16,55 +16,114 @@ package policy
+}
+
+func mergeL4Port(r api.PortRule, p api.PortProtocol, proto string, resMap L4PolicyMap) int {
+	fmt := fmt.Sprintf("%s:%d", proto, p.Port)
```
why not `%s/%d` instead?

tgraf replied:
Either is fine, it doesn't matter. Never exposed. Will change this if you look this better.

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,259 @@
+// - All members of this structure are optional. If omitted or empty, the
+//   member will have no effect on the rule.
+// - All members of this structure are evaluated independently, i.e. L4 ports
+//   allowed with ToPorts do not depend on a match of the FromEndpoints in the
```
Please update the comment; There is no FromEndpoints in the EgressRule.

tgraf replied:
Fixed

ashwinp commented on pkg/policy/api/rule.go:
```diff
@@ -0,0 +1,259 @@
+//
+// - All members of this structure are optional. If omitted or empty, the
+//   member will have no effect on the rule.
+// - All members of this structure are evaluated independently, i.e. L4 ports
```
"All members of this structure are evaluated independently". This would imply that as a user, I cannot enforce the following:
- Allow Ingress only from container labeled A (FromEndpoint=A) to port 80.
- Allow Ingress only from container labeled B (FromEndpoint=B) to port 8080.

B will be allowed to connect to port 80 in addition to port 8080. Similarly, A will be allowed to connect to port 8080 in addition to port 80. Right?

tgraf replied:
Correct. Two reasons:
- The BPF datapath is currently not capable of enforcing this. L3 endpoints are whitelisted via BPF map. The per endpoint L4 map is another BPF map. Referring to a 2nd BPF map from a 1st BPF map value was only just recently added so we will start supporting this with more recent kernels.
- k8s NetworkPolicy also defines this separate notion and we need to able to map k8s NetworkPolicy to our policy.

The idea is to add a field later on which introduces L4 policies which depend on a specific L3 rule. This could be done by adding a new `ToPorts` inside `FromEndpoints` or to add a flag to `ToPorts` which allows to tie the `ToPorts` to `FromEndpoints` and `FromCIDR` within the same IngressRule.

tgraf committed:
policy: Adjust code to new simplified policy API

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Adapt /policy API

- Removes the {path} parameter from the API
- Introduces search & delete by labels

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
labels: New ParseLabelArray()

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Simplified policy API

- Replacement of tree structure with list of rules. Within the list, all
  rules enjoy the same priority
- Separation of concerns will be addresses via a different concept
  introduced later. Users without an interest in separation of concern
  are not introduce to unnecessary complexity
- Removal of deny rules. Require rules provide a great baseline and all
  use cases so far could be addressed with require rules without
  requiring implicit denies.
- Removal of tree allows to drastically simplify label handling as
  labels are now always absolute. No need for a label to have a path.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Better debug messages during policy regeneration

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
Documentation: Remove outdated policy model

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
tests: Adjust tests to new policy

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
tests/monitor: Longer wait times to gather notifications

Fixes sporadic testsuite failures

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
tests: Allow calling monitor_stop on stopped monitor

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm reviewed (approved):
Really awesome work @tgraf !
